sdk/all/src/jmh/java/io/opentelemetry/sdk/trace/SpanPipelineBenchmark.java

```diff
@@ -5,96 +5,123 @@
+import org.openjdk.jmh.annotations.Mode;
 import org.openjdk.jmh.annotations.OutputTimeUnit;
 import org.openjdk.jmh.annotations.Scope;
 import org.openjdk.jmh.annotations.Setup;
 import org.openjdk.jmh.annotations.State;
 import org.openjdk.jmh.annotations.Threads;
 import org.openjdk.jmh.annotations.Warmup;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.MountableFile;
 
-@State(Scope.Benchmark)
 public class SpanPipelineBenchmark {
+  private SpanPipelineBenchmark() {}
 
-  private static final AttributeKey<String> OPERATION_KEY = stringKey("operation");
-  private static final AttributeKey<Long> LONG_ATTRIBUTE_KEY = longKey("longAttribute");
-  private static final AttributeKey<String> STRING_ATTRIBUTE_KEY = stringKey("stringAttribute");
-  private static final AttributeKey<Double> DOUBLE_ATTRIBUTE_KEY = doubleKey("doubleAttribute");
-  private static final AttributeKey<Boolean> BOOLEAN_ATTRIBUTE_KEY = booleanKey("booleanAttribute");
-  private final Tracer tracer = OpenTelemetry.getGlobalTracerProvider().get("benchmarkTracer");
-
-  @Setup(Level.Trial)
-  public final void setup() {
-    SpanExporter exporter = new NoOpSpanExporter();
-    OpenTelemetrySdk.getGlobalTracerManagement()
-        .addSpanProcessor(SimpleSpanProcessor.builder(exporter).build());
-  }
+  @State(Scope.Benchmark)
+  public abstract static class AbstractProcessorBenchmark {
+    private static final DockerImageName OTLP_COLLECTOR_IMAGE =
+        DockerImageName.parse("otel/opentelemetry-collector-dev:latest");
+    private static final int EXPOSED_PORT = 5678;
+    private static final int HEALTH_CHECK_PORT = 13133;
+    private SpanBuilderSdk spanBuilderSdk;
+    protected abstract SpanProcessor getSpanProcessor(String collectorAddress);
+    protected abstract void runThePipeline();
 
-  @Benchmark
-  @Threads(value = 5)
-  @Fork(1)
-  @Warmup(iterations = 5, time = 1)
-  @Measurement(iterations = 5, time = 1)
-  @OutputTimeUnit(TimeUnit.MILLISECONDS)
-  public void runThePipeline_05Threads() {
-    doWork();
-  }
+    protected void doWork() {
+      Span span = spanBuilderSdk.startSpan();
+      for (int i = 0; i < 10; i++) {
+        span.setAttribute("benchmarkAttribute_" + i, "benchmarkAttrValue_" + i);
+      }
+      span.end();
+    }
+
+    @Setup(Level.Trial)
+    public void setup() {
+      // Configuring the collector test-container
+      GenericContainer<?> collector =
+          new GenericContainer<>(OTLP_COLLECTOR_IMAGE)
+              .withExposedPorts(EXPOSED_PORT, HEALTH_CHECK_PORT)
+              .waitingFor(Wait.forHttp("/").forPort(HEALTH_CHECK_PORT))
+              .withCopyFileToContainer(
+                  MountableFile.forClasspathResource("/otel.yaml"), "/etc/otel.yaml")
+              .withCommand("--config /etc/otel.yaml");
+
+      collector.start();
```

Comment:
While ryuk cleans things up usually, let's go ahead and add `collector.stop()` in a teardown

Reply:
An issue with doing `collector.stop()` in the teardown method is that the collector is shutdown before the span processor is able to export all the spans from the last benchmark iteration. For BSP, the spans can be flushed out before stopping the collector container but SSP doesn't offer any such functionality.
Letting ryuk take care of the collector testContainer for us is probably the best way right now.